File: src/java/magma/io/Path.java
package magma.io;

import magma.java.JavaSet;
import magma.option.Option;
import magma.result.Result;
import magma.stream.Stream;

public interface Path {
    Stream<Path> streamNames();

    boolean exists();

    Result<JavaSet<Path>, Error> walk();

    Result<String, Error> readString();

    Option<Error> writeString(String output);

    Option<Error> createDirectories();

    Path relativize(Path child);

    Option<Path> findParent();

    Path resolve(String segment);

    java.nio.file.Path unwrap();

    boolean isRegularFile();

    Path findFileName();
}

File: src/c/magma/api/io/Path.c
#include "./Path.h"
struct Path{
	boolean isExists();
	Stream<Path> stream();
	Option<IOException> writeString(String output);
	Option<IOException> createAsDirectories();
	Result<String, IOException> readStrings();
	Result<JavaSet<Path>, IOException> walkWrapped();
	java.nio.file.Path unwrap();
	Path relativize(Path child);
	Option<Path> findParent();
	int getNameCount();
	Option<Path> getName(int index);
	String format();
	Path getFileName();
	Path resolvePath(Path child);
	Path resolveChild(String child);
}
